### ISSUE TITLE
redis/resque: on 2022-09-15, why did resque seem to lose track of worker counts and drop the ibm part audit queue and the (named) failure queues, while keeping all the rest of the queue contents?

### ISSUE THREAD

**Issue by jmartin-sul** [sul-dlss/preservation_catalog#1942]
tony z saw no errors in redis logs, but redis memory usage had been quite high for the prior day or two.  see https://stanfordlib.slack.com/archives/C09M7P91R/p1663266607916699

**Comment by jmartin-sul** [comment]
TLDR: @jmartin-sul thinks that Redis was behaving as configured.

Gory details from looking into the issue:

* Couldn't figure out why the ibm part audit queue contents and the empty failure queues disappeared, while the other two part audit queues and the c2m queue stayed populated
  * This might be explained by https://redis.io/docs/manual/eviction/
    * Looking into "eviction" specifically came out of running this by Rob and Kam while we were waiting on puppet runs during the pres cat re-IP; I described the unexpected behavior of older queue entries disappearing, which made it seem like we were not experiencing a crash and subsequent recovery from an old snapshot with resultant loss of recent data, since it was the older data that was lost; can't remember who said "eviction" first, but Rob did have a vague memory that there might be redis config around that, and indeed there is ("maybe redis has a size limit on what it can hold, and adding new stuff pushes out old stuff?")
    * You can set `maxmemory` threshold and pick the eviction behavior when the addition of new data would cause Redis' memory usage to exceed the specified threshold (specifying `maxmemory` of 0 sets to unlimited on a 64 bit system, implicit limit is 3 GB on 32 bit).  Eviction behavior controlled by `maxmemory_policy` setting.
    * pres cat prod redis seems to set the limit to 5GB with a `maxmemory_policy` of `allkeys-lru`
      * per redis docs: `allkeys-lru` "Keeps most recently used keys; removes least recently used (LRU) keys"
        * This and the lack of overt errors from Redis make it seem like we saw eviction per the configured behavior, rather than Redis crashing and reading from an out of date snapshot; `c2m`, the queue with the newer values was kept, `part_audit_ibm_us_south` and the error queues presumably got dropped because their contents were older
      * pres cat Redis config: https://github.com/sul-dlss/puppet/blob/1dec8838da60641ebfc9f66c5a343ce451d86161/hieradata/node/preservation-catalog-prod-05.stanford.edu.eyaml#L48-L49
* Follow-on work: can we adjust things so that:
  * pres cat is given a higher memory limit in redis settings (10 GB?).  It seems most apps have 5 GB, but pres cat (due to audits) may at times have deeper queues than most other Redis-using applications
  * Whatever the memory limit is, the critical Nagios warning is set to be far enough below the memory limit to give us ample warning (would 1 GB be sufficient?).
    * But... what can we do when approaching the memory limit?  How does a random ops or infra person know the implication of a critical warning?
    * Can the Nagios warning mention that things may get evicted or not added depending on the app's specific policy?  Maybe also suggests manually dumping an audit queue that can be rebuilt later with relative ease, and/or stopping something that's queueing too aggressively considering existing queue depths?
  * Is it possible for a `noeviction` policy to throw an error that'd cause a Ruby exception and HB alert if something can't be queued?  Is that even desirable?  warrants discussion.
    * This seems like that is the default/only behavior for `noeviction`? -- "Redis can return errors for commands that could result in more memory being used, or it can evict some old data to return back to the specified limit every time new data is added." (Redis manual entry on eviction)
      * https://joshtronic.com/2020/09/27/how-to-set-redis-max-memory-usage/ -- "Redis will be unable to write any new keys/values and will error... is default behavior"
* Related unexpected behavior: for a while after the hiccup, we only saw a `failed` queue in Redis web console, instead of the individually named failure queues we'd seen before.
  * Seems like individual failure queues not showing up even though some `zip_endpoint_events` jobs failed is an artifact of Resque web's slightly buggy logic for deciding whether to show multiple failure queues
    * https://github.com/resque/resque/blob/v1.27.4/lib/resque/server/views/queues.erb#L43-L55 --  `if failed_multiple_queues?`
    * https://github.com/resque/resque/blob/v1.27.4/lib/resque/server/helpers.rb#L10-L13  -- crux of `failed_multiple_queues?` is `Resque::Failure.queues.size > 1`
    * But since the failure queues disappeared, even though `Resque::Failure.backend == Resque::Failure::RedisMultiQueue` and `Resque::Failure.backend.queues == ["zip_endpoint_events_failed"]`, it still looks like a single failure queue setup for the overly simple display logic
    * We now have display of multiple failure queues back, since we've had a small number of failures from multiple queues (as of 2022-10-04, `checksum_validation_failed`, `m2c_failed`, `part_audit_aws_s3_east_1_failed`, `zip_endpoint_events_failed`)

**Comment by jmartin-sul** [comment]
i believe this ticket was just about investigating why this happened.  i think all the needed follow-on work has been captured in a new ops tasks ticket.  closing this one.
